packages/jupyterlab-manager/src/renderer.ts

```diff
@@ -16,51 +20,84 @@ import {
 import {
     WidgetManager
 } from './manager';
+import { WidgetModel } from '@jupyter-widgets/base';
 
 /**
  * A renderer for widgets.
  */
 export
 class WidgetRenderer extends Panel implements IRenderMime.IRenderer, IDisposable {
-    constructor(options: IRenderMime.IRendererOptions, manager: WidgetManager) {
+    constructor(options: IRenderMime.IRendererOptions, manager?: WidgetManager) {
         super();
         this.mimeType = options.mimeType;
-        this._manager = manager;
+        if (manager) {
+          this.manager = manager;
+        }
     }
 
+  /**
+   * The widget manager.
+   */
+  set manager(value: WidgetManager) {
+    value.restored.connect(this._rerender, this);
+    this._manager.resolve(value);
+  }
+
   async renderModel(model: IRenderMime.IMimeModel) {
     const source: any = model.data[this.mimeType];
-
+    const manager = await this._manager.promise;
     // If there is no model id, the view was removed, so hide the node.
     if (source.model_id === '') {
       this.hide();
       return Promise.resolve();
     }
 
-    const modelPromise = this._manager.get_model(source.model_id);
-    if (modelPromise) {
-      try {
-        let wModel = await modelPromise;
-        let widget = await this._manager.display_model(void 0, wModel, void 0);
-        this.addWidget(widget);
-
-        // If the widget is disposed, hide this container and make sure we
-        // change the output model to reflect the view was closed.
-        widget.disposed.connect(() => {
-          this.hide();
-          source.model_id = '';
-        });
-      } catch (err) {
-        console.log('Error displaying widget');
-        console.log(err);
-        this.node.textContent = 'Error displaying widget';
-        this.addClass('jupyter-widgets');
-      }
-    } else {
-      this.node.textContent = 'Error creating widget: could not find model';
+    let wModel: WidgetModel;
+    try {
+      wModel = await manager.get_model(source.model_id);
+    } catch (err) {
+      console.log('Error getting widget state');
+      console.log(err);
+
+      // Let's be optimistic, and hope the widget state will come later.
+      this.node.textContent = 'Loading widget...';
       this.addClass('jupyter-widgets');
-      return Promise.resolve();
+
+      // If we haven't rerendered in 5 seconds with the right model, display an
+      // error message to the user.
+      setTimeout(() => {
+        if (this._rerenderMimeModel) {
+          this.node.textContent = 'Error displaying widget: model not found';
```

Comment:
Maybe we should tweak this error message to indicate that, while no actual error has occurred, getting the widget state is taking a lot longer than expected.

Reply:
I pushed a commit that tries to make the error messages more intelligent, based on whether the manager has restored its status or not.